src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWallet.kt

```diff
@@ -0,0 +1,188 @@
+            val witness = outPoint2Address[txIn.outPoint]?.let { address ->
+                addresses[address]?.find { it.outPoint == txIn.outPoint }?.let { utxo ->
+                    privateKeys[address]?.let { privateKey ->
+                        // mind this: the pubkey script used for signing is not the prevout pubscript (which is just a push
+                        // of the pubkey hash), but the actual script that is evaluated by the script engine, in this case a PAY2PKH script
+                        val publicKey = privateKey.publicKey()
+                        val pubKeyScript = Script.pay2pkh(publicKey)
+                        val sig = Transaction.signInput(
+                            tx,
+                            index,
+                            pubKeyScript,
+                            SigHash.SIGHASH_ALL,
+                            utxo.value.sat,
+                            SigVersion.SIGVERSION_WITNESS_V0,
+                            privateKey
+                        )
+                        Script.witnessPay2wpkh(publicKey, sig.byteVector())
+                    }
+                }
+            }
+            when (witness) {
+                is ScriptWitness ->
+                    // update the signature for the corresponding input
+                    wipTx.updateWitness(index, witness)
+                else ->
+                    // we don't know how to sign this input
+                    wipTx
+            }
+        }
+    }
+}
+
+private sealed interface WalletCommand {
+    companion object {
+        object ElectrumConnected : WalletCommand
+        data class ElectrumNotification(val msg: ElectrumResponse) : WalletCommand
+        data class AddAddress(val bitcoinAddress: String, val privateKey: PrivateKey?) : WalletCommand
+    }
+}
+
+
+/**
+ * A very simple wallet that only watches one address and publishes its utxos.
+ */
+class ElectrumMiniWallet(
+    val chainHash: ByteVector32,
+    private val client: ElectrumClient,
+    private val scope: CoroutineScope
+) : CoroutineScope by scope {
+
+    // state flow with the current balance
+    private val _walletStateFlow = MutableStateFlow(WalletState(emptyMap(), emptyMap(), emptyMap()))
+    val walletStateFlow get() = _walletStateFlow.asStateFlow()
+
+    // all currently watched script hashes and their corresponding bitcoin address
+    private var scriptHashes: Map<ByteVector32, String> = emptyMap()
+
+    // the mailbox of this "actor"
+    private val mailbox: Channel<WalletCommand> = Channel(Channel.BUFFERED)
+
+    internal fun addAddress(bitcoinAddress: String, privateKey: PrivateKey? = null) {
+        launch {
+            mailbox.send(WalletCommand.Companion.AddAddress(bitcoinAddress, privateKey))
+        }
+    }
+
+    fun addPay2wpkhAddress(privateKey: PrivateKey) {
+        launch {
+            val bitcoinAddress = Bitcoin.computeP2WpkhAddress(privateKey.publicKey(), Block.TestnetGenesisBlock.hash)
```

Comment:
We'll need to make the `chainHash` an argument of this function, right? Why not do it now?

Reply:
Damn, fixed in 865eda8.